File: src/main/java/com/nao4j/subtrol/provider/CurrencyRatesProviderFixerApi.java
package com.nao4j.subtrol.provider;

import com.nao4j.subtrol.document.CurrencyRates;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Repository;
import org.springframework.web.client.RestTemplate;

import java.util.Optional;

@Repository
@RequiredArgsConstructor
public class CurrencyRatesProviderFixerApi implements CurrencyRatesProvider {

    private static final String FIXER_LASTES_URI = "http://data.fixer.io/api/latest?access_key={accessKey}";

    private final RestTemplate restTemplate;
    @Value("${api.fixer.key}")
    private String accessKey;

    @Override
    public Optional<CurrencyRates> latest() {
        return Optional.ofNullable(restTemplate.getForEntity(FIXER_LASTES_URI, CurrencyRates.class, accessKey))
                .map(ResponseEntity::getBody);
    }

}

File: src/main/kotlin/com/nao4j/subtrol/provider/CurrencyRatesProviderFixerApi.kt
package com.nao4j.subtrol.provider

import com.nao4j.subtrol.model.CurrencyRates
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Repository
import org.springframework.web.client.RestTemplate

@Repository
class CurrencyRatesProviderFixerApi(
    private val restTemplate: RestTemplate,
    @Value("\${api.fixer.key}") private val accessKey: String
) : CurrencyRatesProvider {

    override fun latest(): CurrencyRates? = restTemplate.getForEntity<CurrencyRates>(
        "http://data.fixer.io/currencyRatesProvider/latest?access_key={accessKey}",
        CurrencyRates::class.java,
        accessKey
    ).body

}
